lsp-mode.el

```diff
@@ -1877,6 +1897,101 @@ WORKSPACE is the workspace that contains the progress token."
+                                                lsp-seq-first
+                                                (gethash "title")
+                                                (replace-regexp-in-string "[\n\t ]+" " "))
+                                           'face 'lsp-modeline-code-actions-face))
+          (single-action? (= (length actions) 1))
+          (string (if single-action?
+                      (format " %s %s " icon first-action-string)
+                    (format " %s %s %s " icon first-action-string
+                            (propertize (format "(%d more)" (seq-length actions))
+                                        'display `((height 0.9))
+                                        'face 'lsp-modeline-code-actions-face)))))
+    (propertize string
+                'help-echo (concat "Apply code actions (s-l a a)\nmouse-1: "
+                                   (if single-action?
+                                       first-action-string
+                                     "select from multiple code actions"))
+                'mouse-face 'mode-line-highlight
+                'local-map (make-mode-line-mouse-map
+                            'mouse-1 (lambda ()
+                                       (interactive)
+                                       (if single-action?
+                                           (lsp-execute-code-action (lsp-seq-first actions))
+                                         (call-interactively #'lsp-execute-code-action)))))))
+
+(defun lsp-modeline--update-code-actions (actions)
+  "Update modeline with new code ACTIONS."
+  (when lsp-modeline-code-actions-kind-regex
+    (setq actions (seq-filter (-lambda ((&hash "kind"))
+                                (or (not kind)
+                                    (s-match lsp-modeline-code-actions-kind-regex kind)))
+                              actions)))
+  (if (seq-empty-p actions)
+      (setq-local global-mode-string (remove '(t (:eval lsp--modeline-code-actions-string)) global-mode-string))
+    (progn
+      (setq lsp--modeline-code-actions-string (lsp--modeline-build-code-actions-string actions))
+      (add-to-list 'global-mode-string '(t (:eval lsp--modeline-code-actions-string))))))
+
+(defun lsp--modeline-request-code-actions ()
+  "Requests for code actions to update modeline."
+  (lsp-request-async
+   "textDocument/codeAction"
+   (lsp--text-document-code-action-params)
+   #'lsp-modeline--update-code-actions
+   :mode 'alive
+   :cancel-token :lsp-modeline-code-actions))
+
+(defvar-local lsp--modeline-code-actions--timer nil)
+
+(defun lsp--modeline-check-code-actions ()
+  "Check for code actions showing it on modeline."
+  (if t
+      (progn
+        (when lsp--modeline-code-actions--timer
+          (cancel-timer lsp--modeline-code-actions--timer))
+        (let ((buf (current-buffer)))
+          (setq lsp--modeline-code-actions--timer
+                (run-with-idle-timer lsp-modeline-code-actions-delay
```

Comment:
Can we reuse diagnostic information here?
Or only if current line has diagnostics information?
From the definition of `CodeAction` and related, it has
```

	/**
	 * The diagnostics that this code action resolves.
	 */
	diagnostics?: Diagnostic[];
```
so it is tied to diagnostics information.